zingolib/src/wallet/transaction_records_by_id.rs

```diff
@@ -211,95 +211,123 @@ impl TransactionRecordsById {
+    /// found in the wallet, indicating that all shielded spends have not yet been synced.
+    /// Also returns this error
+    /// if the transaction record contains outgoing_tx_data but no spends are found.
+    /// If a transparent spend has not yet been synced, the fee will be incorrect and return
+    /// [`crate::wallet::error::FeeError::FeeUnderflow`] if an underflow occurs.
+    /// The tracking of transparent spends will be improved on the next internal wallet version.
+    pub fn calculate_transaction_fee(
+        &self,
+        query_record: &TransactionRecord,
+    ) -> Result<u64, FeeError> {
+        let sapling_spend_value: u64 = self
+            .get_sapling_notes_spent_in_tx(query_record)?
+            .iter()
+            .map(|&note| note.value())
+            .sum();
+        let orchard_spend_value: u64 = self
+            .get_orchard_notes_spent_in_tx(query_record)?
+            .iter()
+            .map(|&note| note.value())
+            .sum();
+        let total_trans_spent = query_record.total_transparent_value_spent;
+        let total_spend_value = query_record.total_transparent_value_spent
+            + total_trans_spent
             + sapling_spend_value
             + orchard_spend_value;
 
         if total_spend_value == 0 {
-            if transaction_record.value_outgoing() == 0 {
+            if query_record.value_outgoing() == 0 {
                 return Err(FeeError::ReceivedTransaction);
             } else {
-                return Err(FeeError::SpendNotFound);
+                return Err(FeeError::OutgoingWithoutSpends(
+                    query_record.outgoing_tx_data.to_vec(),
+                ));
             }
         }
 
-        let transparent_output_value: u64 = transaction_record
+        let transparent_output_value: u64 = query_record
             .transparent_outputs()
             .iter()
             .map(|note| note.value())
             .sum();
-        let sapling_output_value: u64 = transaction_record
+        let sapling_output_value: u64 = query_record
             .sapling_notes()
             .iter()
             .map(|note| note.value())
             .sum();
-        let orchard_output_value: u64 = transaction_record
+        let orchard_output_value: u64 = query_record
             .orchard_notes()
             .iter()
             .map(|note| note.value())
             .sum();
 
-        let total_output_value = transaction_record.value_outgoing()
+        let total_output_value = query_record.value_outgoing()
             + transparent_output_value
             + sapling_output_value
             + orchard_output_value;
 
         if total_spend_value >= total_output_value {
             Ok(total_spend_value - total_output_value)
         } else {
-            Err(FeeError::FeeUnderflow)
+            Err(FeeError::FeeUnderflow((
+                query_record.total_value_spent(),
+                query_record.total_value_received(),
```

Comment:
please replace with `total_spend_value` and `total_output value`

Reply:
Done.